wagtail_streamfield_migration_toolkit/autodetect/streamchangedetector.py

```diff
@@ -99,31 +100,46 @@ def find_renamed_or_removed_defs(
             old_child_def = old_child_defs[old_child_name]
             is_child_mapped = False
 
-            # For now assume that same name means same block (i.e. the old block wasn't deleted
-            # and a new block with the same name added)
+            # Find out if the block maps to one of the new only children. If it maps, that
+            # would mean that the block has been renamed
+
+            comparer = block_def_comparer_registry.get_block_def_comparer_for_instance(
+                old_child_def
+            )
             if old_child_name in new_child_names:
-                new_child_def = new_child_defs[old_child_name]
-                # recursion call
-                self.find_renamed_or_removed_defs(
-                    old_child_def,
-                    new_child_def,
-                    old_child_path,
+                new_child_name = old_child_name
+                new_child_def = new_child_defs[new_child_name]
+                new_child_path = parent_path + path_suffix + new_child_name
+
+                similarity_score = comparer.compare(
+                    old_def=old_child_def,
+                    old_name=old_child_name,
+                    new_def=new_child_def,
+                    new_name=new_child_name,
                 )
-                is_child_mapped = True
-                # TODO add another threshold for cases where we can be sure that the blocks are the
-                # same without asking the user.
-
-            else:
-                comparer = (
-                    block_def_comparer_registry.get_block_def_comparer_for_instance(
-                        old_child_def
+                if similarity_score >= CONFIDENT_SIMILARITY_THRESHOLD:
+
+                    # recursion call
+                    self.find_renamed_or_removed_defs(
+                        old_child_def,
+                        new_child_def,
+                        old_child_path,
                     )
-                )
+                    is_child_mapped = True
 
-                # Find out if the block maps to one of the new only children. If it maps, that
-                # would mean that the block has been renamed
+                elif similarity_score >= VERIFYING_SIMILARITY_THRESHOLD:
+                    if self.questioner.ask_block_not_changed(old_path=old_child_path):
```

Comment:
Maybe in future we should see if we can separate the questioner from this recursive comparison process. Having it here means it adds workload when working on the comparison process, and adds complexity to our tests. Walking down the tree and comparing block definitions is what I would consider a low level operation (relative to this library), and interacting with the user is very high level.